pkg/helpers/tokencmd/negotiator_gssapi.go

```diff
@@ -4,195 +4,91 @@
-	lib, err := g.loadLib()
-	if err != nil {
-		return nil, err
-	}
-
-	// Initialize our context if we haven't already
-	if g.ctx == nil {
-
-		if len(g.principalName) > 0 {
-			// Get credentials for a specific principal
-			klog.V(5).Infof("acquiring credentials for principal name %s", g.principalName)
-			credBuffer, err := lib.MakeBufferString(g.principalName)
-			if err != nil {
-				return nil, convertGSSAPIError(err)
-			}
-			defer credBuffer.Release()
-
-			credName, err := credBuffer.Name(lib.GSS_KRB5_NT_PRINCIPAL_NAME)
-			if err != nil {
-				return nil, convertGSSAPIError(err)
-			}
-			defer credName.Release()
-
-			cred, _, _, err := lib.AcquireCred(credName, time.Duration(0), lib.GSS_C_NO_OID_SET, gssapi.GSS_C_INITIATE)
-			if err != nil {
-				klog.V(5).Infof("AcquireCred returned error: %v", err)
-				return nil, convertGSSAPIError(err)
-			}
-			g.cred = cred
-		} else {
-			// otherwise, express no opinion about the credentials and let gssapi decide
-			g.cred = lib.GSS_C_NO_CREDENTIAL
+	if g.client == nil {
+		krb5ConfLocation := krb5Config
+		if os.Getenv(krb5ConfigEnvKey) != "" {
+			krb5ConfLocation = os.Getenv(krb5ConfigEnvKey)
 		}
 
-		serviceName, err := getServiceName('@', requestURL)
+		config, err := krb5conf.Load(krb5ConfLocation)
 		if err != nil {
 			return nil, err
 		}
 
-		klog.V(5).Infof("importing service name %s", serviceName)
-		nameBuf, err := lib.MakeBufferString(serviceName)
+		krb5KeytabLocation := krb5Keytab
+		if os.Getenv(krb5KeytabEnvKey) != "" {
+			krb5KeytabLocation = os.Getenv(krb5KeytabEnvKey)
+		}
+		keytab, err := krb5kt.Load(krb5KeytabLocation)
 		if err != nil {
-			return nil, convertGSSAPIError(err)
+			return nil, err
 		}
-		defer nameBuf.Release()
 
-		name, err := nameBuf.Name(lib.GSS_C_NT_HOSTBASED_SERVICE)
+		g.client = krb5client.NewWithKeytab(g.principalName, config.LibDefaults.DefaultRealm, keytab, config)
+
+		err = g.client.Login()
 		if err != nil {
-			return nil, convertGSSAPIError(err)
+			return nil, err
 		}
-		g.name = name
-		g.ctx = lib.GSS_C_NO_CONTEXT
 	}
 
-	incomingTokenBuffer, err := lib.MakeBufferBytes(challengeToken)
+	servicePrincipleName, err := getServiceName('/', requestURL)
 	if err != nil {
-		return nil, convertGSSAPIError(err)
+		return nil, err
 	}
-	defer incomingTokenBuffer.Release()
-
-	var outgoingToken *gssapi.Buffer
-	g.ctx, _, outgoingToken, _, _, err = lib.InitSecContext(g.cred, g.ctx, g.name, lib.GSS_C_NO_OID, g.flags, time.Duration(0), lib.GSS_C_NO_CHANNEL_BINDINGS, incomingTokenBuffer)
-	defer outgoingToken.Release()
-
-	switch err {
-	case nil:
-		klog.V(5).Infof("InitSecContext returned GSS_S_COMPLETE")
-		g.complete = true
-		return outgoingToken.Bytes(), nil
-	case gssapi.ErrContinueNeeded:
-		klog.V(5).Infof("InitSecContext returned GSS_S_CONTINUE_NEEDED")
+
+	_, key, err := g.client.GetServiceTicket(servicePrincipleName)
+	if err != nil {
+		g.client = nil
 		g.complete = false
```

Comment:
I said complete to false in case of errors which is another difference from the referenced PR